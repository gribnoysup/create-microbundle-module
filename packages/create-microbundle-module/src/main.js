import fs from 'fs';
import path from 'path';
import programm from 'commander';
import createModule from './create-module';

let moduleDirectory;

const { name, version } = fs.readFileSync(
  path.resolve(__dirname, '..', 'package.json')
);

programm
  .name(name)
  .version(version)
  .arguments('<module-directory>')
  .option(
    '-t, --target <target>',
    'Your module target environment: web, node (default web)',
    /^(web|node)$/i,
    'web'
  )
  .option(
    '-s, --scripts-version <alternative-version>',
    'use a non-standard version of microbundle-module-scripts',
    'microbundle-module-scripts'
  )
  // TODO: We will need additional bootstrapping for TS (or Reason, when
  // it lands https://github.com/developit/microbundle/pull/142)
  // .option(
  //   '-l, --language',
  //   'Create JavaScript, TypeScript or Reason module',
  //   /^(javascript|typescript|reason)$/i,
  //   'javascript'
  // )
  .action(dir => (moduleDirectory = dir));

programm.parse(process.argv);

createModule(moduleDirectory, {
  target: programm.target,
  scriptsVersion: programm.scriptsVersion,
});
