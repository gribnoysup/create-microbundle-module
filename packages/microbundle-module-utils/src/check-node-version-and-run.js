import child_process from 'child_process';
import pleaseUpgradeNode from 'please-upgrade-node';
import pkgUp from 'pkg-up';
import chalk from 'chalk';
import symbols from 'log-symbols';

export default function checkNodeVersionAndRun(modulePath) {
  return pkgUp(modulePath).then(function(packagePath) {
    if (packagePath !== null) {
      var pkg = require(packagePath);

      pleaseUpgradeNode(pkg, {
        message: function message(requiredVersion) {
          const version = process.versions.node;
          const message =
            `${symbols.warning}  You are running Node ${version}.\n` +
            `"${pkg.name}" requires Node ${requiredVersion} or higher. ` +
            `Please update your version of Node.`;

          return chalk.yellow(message);
        },
      });
    }

    return child_process.fork(modulePath, process.argv.slice(2));
  });
}
