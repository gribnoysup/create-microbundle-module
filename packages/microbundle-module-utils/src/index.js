export { default as logger } from './logger';
export {
  default as checkNodeVersionAndRun,
} from './check-node-version-and-run';
