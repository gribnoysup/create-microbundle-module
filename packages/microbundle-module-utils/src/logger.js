import ora from 'ora';
import chalk from 'chalk';
import symbols from 'log-symbols';

class Logger {
  constructor(name) {
    this.name = name;
    this.spinner = ora();
  }

  __stringifyArgsWithPrefix(args) {
    const str = args
      .map(a => a.toString())
      .join(' ')
      .trim();

    return str
      ? `${this.name ? chalk.dim(`[${this.name}]`) : ''} ${str}`.trim()
      : void 0;
  }

  child(name) {
    return new Logger(`${this.name}/${name}`);
  }

  fresh(name) {
    return new Logger(name);
  }

  log(...args) {
    const message = this.__stringifyArgsWithPrefix(args);

    if (message) {
      // eslint-disable-next-line no-console
      console.log(message);
    } else {
      // eslint-disable-next-line no-console
      console.log();
    }
  }

  info(...args) {
    const message = this.__stringifyArgsWithPrefix(args);

    if (this.spinner.isSpinning) {
      this.spinner.info(message);
    } else {
      // eslint-disable-next-line no-console
      console.log(symbols.info, message);
    }
  }

  success(...args) {
    const message = this.__stringifyArgsWithPrefix(args);

    if (this.spinner.isSpinning) {
      this.spinner.succeed(message);
    } else {
      // eslint-disable-next-line no-console
      console.log(symbols.success, message);
    }
  }

  warning(...args) {
    const message = this.__stringifyArgsWithPrefix(args);

    if (this.spinner.isSpinning) {
      this.spinner.warn(message);
    } else {
      // eslint-disable-next-line no-console
      console.warn(symbols.warning, message);
    }
  }

  error(...args) {
    const message = this.__stringifyArgsWithPrefix(args);

    if (this.spinner.isSpinning) {
      this.spinner.fail(message);
    } else {
      // eslint-disable-next-line no-console
      console.error(symbols.error, message);
    }
  }

  start(...args) {
    this.spinner.start(this.__stringifyArgsWithPrefix(args));
  }

  stop() {
    this.spinner.stop();
  }
}

const defaultLogger = new Logger('microbundle-module');

export default defaultLogger;
