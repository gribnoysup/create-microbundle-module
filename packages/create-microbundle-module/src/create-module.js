import path from 'path';
import programm from 'commander';
import fs from 'fs-extra';
import execa from 'execa';
import Git from 'simple-git/promise';
import semver from 'semver';

import camelCase from 'camelcase';
import validateName from 'validate-npm-package-name';

import { logger as defaultLogger } from 'microbundle-module-utils';

const git = Git();
const logger = defaultLogger.fresh();

const replaceModuleName = (str, moduleName) =>
  str
    .replace(/\$moduleName/g, camelCase(moduleName))
    .replace(/\$module-name/g, moduleName);

const getTemplateByName = async templateName => {
  const content = await fs.readFile(
    path.join(__dirname, '..', 'templates', templateName),
    'utf-8'
  );

  return content;
};

const createModule = async (
  moduleDirectory,
  { target, scriptsVersion } = {}
) => {
  if (!moduleDirectory || typeof moduleDirectory !== 'string') {
    logger.warning(
      `Please specify the project directory: $ ${programm.name()} ./my-awesome-module`
    );
    return;
  }

  moduleDirectory = moduleDirectory.trim();

  const modulePath = path.resolve(process.cwd(), moduleDirectory);
  const moduleName = path.basename(modulePath);

  let isGitAvailable = true;

  const { errors, warnings, validForNewPackages: validName } = validateName(
    moduleName
  );

  if (validName === false) {
    const messages = [].concat(errors || [], warnings || []);
    logger.error(`Can't create module with the name "${moduleName}":`);
    messages.forEach(m => logger.error(` - ${m}`));

    process.exitCode = 1;
    return;
  }

  const pkg = JSON.parse(
    // INFO: Template for package.json has 'template' added to the name, because
    // otherwise it throws off npm packaging command and excludes the whole 'templates'
    // folder from being published
    //
    // TODO: Create a small reproducible repo and submit an issue to npm
    replaceModuleName(
      await getTemplateByName('template.package.json'),
      moduleName
    )
  );

  pkg.target = target;

  logger.log();
  logger.log(`Creating a new home for module "${moduleName}" in ${modulePath}`);

  try {
    logger.log();
    logger.start(`Generating initial files`);

    const paths = {
      'entry.js': path.join(modulePath, 'src', `${moduleName}.js`),
      'entry.test.js': path.join(modulePath, 'test', `${moduleName}.test.js`),
      '.gitignore': path.join(modulePath, '.gitignore'),
    };

    for (const key in paths) {
      if (paths.hasOwnProperty(key)) {
        const template = await getTemplateByName(key);
        await fs.outputFile(
          paths[key],
          replaceModuleName(template, moduleName)
        );
      }
    }

    logger.success();
  } catch (error) {
    logger.error('Something went wrong while trying to bootstrap project:');
    logger.error(error);

    process.exitCode = 1;
    return;
  }

  try {
    logger.start('Initializing new git repository');

    await git.cwd(modulePath);
    await git.init();

    const name = await git.raw(['config', 'user.name']);
    const email = await git.raw(['config', 'user.email']);

    if (name || email) {
      pkg.author = [name ? name.trim() : '', email ? `<${email.trim()}>` : '']
        .filter(Boolean)
        .join(' ');
    }

    logger.success();
  } catch (error) {
    isGitAvailable = false;
    logger.warning(
      "Couldn't initialize git repo: git hooks creation and first commit will be skipped"
    );
  }

  try {
    logger.start('Installing external dependencies');

    await fs.writeFile(
      path.join(modulePath, 'package.json'),
      JSON.stringify(pkg, null, 2)
    );

    const semverScriptsVersion = semver.valid(scriptsVersion);
    const dependencies =
      semverScriptsVersion !== null
        ? [`microbundle-module-scripts@${semverScriptsVersion}`]
        : [scriptsVersion.trim()];

    if (!process.env.SKIP_DEPENDENCIES_INSTALL_USE_ONLY_FOR_TESTS) {
      await execa('npm', [
        'install',
        '--save-dev',
        '--save-exact',
        ...dependencies,
      ]);
    }

    logger.success();
  } catch (error) {
    logger.log();
    logger.error('Something went wrong while installing dependencies:');
    logger.error(error);

    return;
  }

  if (isGitAvailable === true) {
    logger.start('Creating initial commit');

    await git.add('*');
    await git.commit(
      'Initial commit (bootstrapped with create-microbundle-module) ðŸš€'
    );

    logger.success();
  }
};

export default createModule;
