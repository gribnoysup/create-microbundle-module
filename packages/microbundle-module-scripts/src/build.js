import fs from 'fs-extra';
import microbundle from 'microbundle';
import { logger as defaultLogger } from 'microbundle-module-utils';

import paths from './paths';
import defaultConfig from './config/microbundle';

const logger = defaultLogger.child('build');

const build = async ({ watch, external, cwd } = {}) => {
  logger.info('Starting module build' + (watch ? ' in watch mode' : ''));

  let pkg;

  if (await fs.exists(paths.appPackageJSONPath)) {
    pkg = await fs.readJSON(paths.appPackageJSONPath);
  }

  const config = {
    ...defaultConfig,
    cwd,
    watch,
    external,
    target: pkg.target,
  };

  try {
    const stats = await microbundle(config);
    logger.log(stats);
  } catch (error) {
    logger.error(error);
    process.exitCode = 1;
  }
};

export default build;
