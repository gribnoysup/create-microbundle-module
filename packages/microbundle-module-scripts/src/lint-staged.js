import execa from 'execa';
import fs from 'fs-extra';
import { logger as defaultLogger } from 'microbundle-module-utils';

import paths from './paths';

const logger = defaultLogger.child('lint-staged');

export const getLintStagedConfigPath = async () => {
  let configPath;

  if ((await fs.exists(paths.appLintStagedConfig)) === true) {
    logger.info('Using lint-staged defined in your module root');
    configPath = paths.appLintStagedConfig;
  } else {
    configPath = paths.ownLintStagedConfig;
  }

  return configPath;
};

const lintStaged = async ({ debug }) => {
  const configPath = await getLintStagedConfigPath();

  return await execa(
    paths.lintStagedBin,
    ['--config', configPath, debug && '--debug'].filter(Boolean),
    { cwd: paths.appPath }
  );
};

export default lintStaged;
