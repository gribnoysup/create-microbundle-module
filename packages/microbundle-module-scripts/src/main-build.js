import commander from 'commander';

import build from './build';
import paths from './paths';

/**
 * TODO: Should we allow more microbundle flags?
 *
 *   -i, --entry      Entry module(s)
 *   -o, --output     Directory to place build files into
 *   -f, --format     Only build specified formats  (default es,cjs,umd)
 * + -w, --watch      Rebuilds on any change  (default false)
 *   --target         Specify your target environment  (default node)
 * + --external       Specify external dependencies, or 'none'
 *   --compress       Compress output using UglifyJS  (default true)
 *   --strict         Enforce undefined global context and add "use strict"
 *   --name           Specify name exposed in UMD builds
 * + --cwd            Use an alternative working directory  (default .)
 *   --sourcemap      Generate source map  (default true)
 */

commander
  .name('build')
  .option('-w, --watch', 'rebuilds on any change', false)
  .option(
    '--external <dependencies>',
    "specify external dependencies, or 'none'",
    'none'
  )
  .option('--cwd <path>', 'use an alternative working directory')
  .parse(process.argv);

build({
  watch: !!commander.watch,
  external: commander.external,
  // INFO: Avoid showing terribly long resolved path to the module
  // in the build --help
  cwd: commander.cwd || paths.appPath,
});
