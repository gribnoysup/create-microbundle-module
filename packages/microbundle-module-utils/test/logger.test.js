import stripAnsi from 'strip-ansi';

import { Logger } from '../src/logger';

describe('logger', () => {
  let originalConsole;
  const logger = new Logger('test');

  const getLastLoggerOutput = type => {
    return stripAnsi(
      global.console[type].mock.calls[
        global.console[type].mock.calls.length - 1
      ].join(' ')
    );
  };

  beforeAll(() => {
    originalConsole = global.console;

    global.console = {
      log: jest.fn(),
      warn: jest.fn(),
      error: jest.fn(),
    };
  });

  afterAll(() => {
    global.console = originalConsole;
  });

  describe('fresh', () => {
    it('should create new instance of logger when called', () => {
      const fresh = logger.fresh();

      expect(fresh === logger).toBeFalsy();
      expect(fresh.constructor === logger.constructor).toBeTruthy();
    });
  });

  describe('child', () => {
    it('should create new instance of logger with concatenated name when called', () => {
      const child = logger.child('child');

      expect(child.name).toBe('test/child');
    });
  });

  describe('log', () => {
    it('should console log input with a prefix', () => {
      logger.log('some', 'text', 'and', 'a', 'number:', 123);

      expect(getLastLoggerOutput('log')).toMatchSnapshot();
    });
  });

  describe('info', () => {
    it('should console log input with a prefix and a info sign', () => {
      logger.info('some', 'text', 'and', 'a', 'number:', 123);

      expect(getLastLoggerOutput('log')).toMatchSnapshot();
    });
  });

  describe('warn', () => {
    it('should console warn input with a prefix and a warn sign', () => {
      logger.warn('some', 'text', 'and', 'a', 'number:', 123);

      expect(getLastLoggerOutput('warn')).toMatchSnapshot();
    });
  });

  describe('error', () => {
    it('should console error input with a prefix and a error sign', () => {
      logger.error('some', 'text', 'and', 'a', 'number:', 123);

      expect(getLastLoggerOutput('error')).toMatchSnapshot();
    });
  });
});
