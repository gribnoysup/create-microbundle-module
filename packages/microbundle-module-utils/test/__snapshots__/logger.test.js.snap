// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`logger error should console error input with a prefix and a error sign 1`] = `"✖ [test] some text and a number: 123"`;

exports[`logger info should console log input with a prefix and a info sign 1`] = `"ℹ [test] some text and a number: 123"`;

exports[`logger log should console log input with a prefix 1`] = `"[test] some text and a number: 123"`;

exports[`logger warn should console warn input with a prefix and a warn sign 1`] = `"⚠ [test] some text and a number: 123"`;
